### ISSUE TITLE
Start using ES6 for development and distribution

### ISSUE THREAD

**Issue by patrickarlt** [Esri/esri-leaflet#525]
Now that ES 2015 (ES6) has been submitted to ECMA General Assembly for approval this seems like an idea time to start thinking about aligning Esri Leaflet with the standard and with the best possible development practices.

First this probably means a few things:

1. Add [Babel](http://babeljs.io/) to the toolchain so we can compile ES 2015 code down to ES 5 code for release and testing.
2. This means that we need a real development environment setup not just the `working` folder that is gitignored

Once those things were done we could take advantage of new features like [arrow functions](http://babeljs.io/docs/learn-es6/#arrows) and [Maps/WeakMap](http://babeljs.io/docs/learn-es6/#map-set-weak-map-weak-set) immediately spring to mind. Of course the biggest change would be using ES 2015 modules.

In addition to ES 2015 modules I still want to support Browserify and AMD based tools without ES 2015 out of the box. There are quite a few variations on the syntaxes for each system so I'll summarize and leave it open for comments:

### Browserify

One require to rule them all - this is how this currently works and is pretty much how everything else works in Browserify land.

```js
var EsriLeaflet = require('esri-leaflet');
```

The maintenance overhead of trying to separate every component into separate modules is probably crazy from a maintenance standpoint because of the whole one repo per module rule. However we could (maybe, might be too crazy) on the commits tagged for release create a whole bunch of CommonJS modules in the root of the repo so you could do:

```js
var FeatureLayer = require('esri-leaflet/Layers/FeatureLayer');
var Query = require('esri-leaflet/Tasks/Query');
```

Or we could use Babel to build all the module 1:1 int he dist folder so the above could also be done with:

```js
var FeatureLayer = require('esri-leaflet/dist/commonjs/Layers/FeatureLayer');
var Query = require('esri-leaflet/dist/commonjs/Tasks/Query');
```

Which has kind of a bad code smell but thats kind of intentional http://blog.izs.me/post/44149270867/why-no-directories-lib-in-node-the-less-snarky but it works.

I need to test the last option but defining `babelify` as a transform in `package.json` might let you do this...

```js
var FeatureLayer = require('esri-leaflet/lib/Layers/FeatureLayer');
var Query = require('esri-leaflet/lib/Tasks/Query');
```

### AMD

The current way is to also do a single require statement for the whole library;

```js
require([
  'esri-leaflet'
], function(EsriLeaflet){
  // do stuff
});
```

But since in AMD land we can map `esri-leaflet` to anywhere we want we can do this to get smaller builds:

```js
require([
  'esri-leaflet/Layers/FeatureLayer',
  'esri-leaflet/Tasks/Query'
], function(FeatureLayer, Query){
  // do stuff
});
```

### JSPM

[JSPM](http://jspm.io/) is a newer package manager designed with ES 2015 modules the solves distribution and packaging problems. it just uses the ES 2015 syntax:

```js
import {FeatureLayer, Query} from 'esri-leaflet';
```

However like most AMD loaders you can also load in individual files from a specified directory. Unlink AMD loaders JSPM lets package owners [declare this directory at the package level](https://github.com/jspm/registry/wiki/Configuring-Packages-for-jspm#directorieslib) so the following syntax also works and will result in smaller builds:

```js
import {FeatureLayer} from 'esri-leaflet/Layers/FeatureLayer';
import {Query} from 'esri-leaflet/Tasks/Query';
```

### Native ES 2015

Identical to the JSPM syntax this means that we will have to distribute Esri Leaflet in ES 2015 form. This would be for people who want to use Esri Leaflet without a package manager but are still using Babel, another compiler,  or use the asynchronous `System.import` syntax:

```js
// sync
import {FeatureLayer, Query} from 'esri-leaflet'; // import everything
import {FeatureLayer} from 'esri-leaflet/layers/FeatureLayer'; // just FeatureLayer and its dependencies
import {Query} from 'esri-leaflet/Tasks/Query'; // just Query and its dependencies

// async
System.import('esri-leaflet').then(function(EsriLeaflet){}) // import everything
System.import('esri-leaflet/layers/FeatureLayer').then(function(FeatureLayer){}); // just FeatureLayer and its dependencies
System.import('esri-leaflet/Tasks/Query').then(function(Query){}); // just Query and its dependencies
```

## Requirements

* Global `L.esri` build. You can download this and host it yourself. This is also what we make available on the CDN.
* CommonJS build for Browserify.
* AMD build for Dojo/RequireJS/other AMD loaders.
* JSPM build (could be CommonJS or ES 2015).
* An ES 2015 build for native ES 2015 support. This should probably also be on the CDN for when browsers support `System.import`.
* Distribute on CDN, GitHub, Bower, NPM and JSPM.

## Workflow

1. Write everything in the ES 2015 syntax
2. To get the `L.esri` build we need a small wrapper file (lets call it `all.js`) that imports everything, and assigns it to `L.esri`. This is what we build and host on the CDN.
3. To support AMD we use Babel with `--modules amd` and distribute it in a special `amd` folder. Users will have to configure the path to `'esri-leaflet'` to point to something like `esri-leaflet/dist/amd`. This is really similar to what you already have to do.
4. To support Browserify we would use Babel with `--modules common` to build our `all.js` file and distribute in a special location like `esri-leaflet/dist/commonjs/esri-leaflet.js` folder. This would be the value of `main` in `package.json` which is used by Browserify when you `require('esri-leaflet')`.
5. At this point JSPM can use same file that Browserify does since it treats everything as CommonJS be default. An alternative to this would be to [point JSPM the folder full of the raw ES 2015 code](https://github.com/jspm/registry/wiki/Configuring-Packages-for-jspm#directorieslib) to allow users use import things on a file-by-file basis like the JSPM example above. You could also override the `main` value to look at the raw source for all.js with the special `jspm` property in `package.json` as [detailed in the JSPM docs](https://github.com/jspm/registry/wiki/Configuring-Packages-for-jspm#prefixing-configuration)
6. At this point we should host the compiled AMD modules and the uncompiled  ES 2015 source code on the CDN to support AMD users not using a package manager and people using `System.import`

## Concerns

* Using Babel will [require use of a polyfill](http://babeljs.io/docs/usage/polyfill/). Luckily we can automatically include this seamlessly in the Browserify/JSPM use cases with the [Babel Runtime](http://babeljs.io/docs/usage/runtime/). However this is a big problem for AMD users or people who want to use the CDN as they now have and additional download to worry about.
* This also means dropping IE 8 support since Babel will assume you have an complete ES 5 environment. This MIGHT be able to work with [es5-shim](https://github.com/es-shims/es5-shim) though.

Any thoughts on this from @odoe @robertd @stdavis @dbouwman @chelm @ngoldman @paulcpederson @nikolaswise @alaframboise @jgravois?

**Comment by paulcpederson** [comment]
@patrickarlt that all looks reasonable. 

One thing I will add though, is you don't necessarily need the polyfill for Babel depending on what featureset you are using. If you don't use things like `Object.assign()`, generators, or `Array.from()` (relied on by the spread operator) you can actually get away with quite a lot sans polyfill. The polyfill isn't included by default, so my advice would be to leave it out until such point that your code breaks.

The official babel polyfill is also quite large as it is a kitchen sink of every polyfill you need for the entire ES6 syntax, so it my be a size concern as well.

**Comment by patrickarlt** [comment]
I would definitely put a ban on generators because Babel needs the entire Regenerator runtime to handle them. Map/WeakMap (if used) needs a polyfill but could be bundled. It would be interesting to see if we could just restrict the amount of ES 6 we use to eliminate the need for a polyfill.

One thing I didn't consider is how would Browserify + Babel or WebPack + Babel would work in this model. The easiest thing to do would be to just require the ES 6 source directly though:

```js
import {FeatureLayer} from 'esri-leaflet/lib/Layers/FeatureLayer';
import {Query} from 'esri-leaflet/lib/Tasks/Query';
```

This works if you define the `babelify` transform in `package.json` but I need to test it first.

**Comment by ngoldman** [comment]
@patrickarlt is Leaflet planning on doing this too? My only reservation would be that Esri-Leaflet should maintain some level of implementation parity with its parent project so as to make it easy to move between the two.

**Comment by stdavis** [comment]
I agree with @ngoldman. Part of the advantage of this project is that it is inline with a mapping library that is already widely accepted and used. Moving away from Leaflet coding patterns and conventions may make it lose some of the great developer experience that it currently has.

But maybe Leaflet is moving in a similar direction? I haven't followed it closely lately.

**Comment by odoe** [comment]
The only real mention of it is here.
https://github.com/Leaflet/Leaflet/issues/3229

But nothing else on the matter.

I'm a little torn. For one, I think there are some definite advantages to it, in particular when using browserify and custom builds. On the other, contributors may be taken back a bit.

But it's not messing with the API, and it's not like ES6 is CoffeeScript or TypeScript, so there is that to consider.

**Comment by dbouwman** [comment]
I tend to agree with @ngoldman - alignment with the core project has a lot of value. 

Additionally, if this complicates life for users of the library, then I'd say it's detrimental to the project, and should be avoided.

**Comment by patrickarlt** [comment]
@ngoldman and @dbouwman I agree that aligning with Leaflet has a lot of value. Implementing changes like these do have quite a bit of value at the build tooling level.

1. I want to get rid of all of the alternate builds http://esri.github.io/esri-leaflet/download/. If you want an alternate build you should use a package manager and a module system. This eliminated a ton of gross stuff from the Gruntfile https://github.com/Esri/esri-leaflet/blob/master/Gruntfile.js#L29-L110
2. I think I can use browserify + babelify to build the general purpose UMD build that gets distributed everywhere. That can remain the default like it is now I think. This will eliminate a whole bunch of grossness from the build process. https://github.com/Esri/esri-leaflet/blob/master/Gruntfile.js#L11-L27
4. Neither of those 2 things will prevent someone from `import {FeatureLayer} from 'esri-leaflet/src/Layers/FeatureLayer';` OR `var FeatureLayer = require('esri-leaflet/src/Layers/FeatureLayer');` So there are still wins to be had for Browserify/JSPM users with no additional cost.
3. If we wanted to keep things simple we could always just not ship an AMD build at this point but the value there would be smaller file sizes for people who are using things like Require JS.

**Comment by patrickarlt** [comment]
@ngoldman @dbouwman I just tested this and it seems like just replacing the build pipeline with Browserify + Babelify would give us some wins.

1. Replaces a ton of silly stuff about concatenating everything in the right order in the Gruntfile
2. Replaces needing to wrap everything in a UMD header/footer
3. We can still export a global `L.esri` and have UMD support in the main distribution file (i.e the CDN one and the one in `main` in package.json, which are the same).
4. If we define the Babelify transform in the `package.json` then people can `require` the source files directly and it will be fine.
5. Since the source files are ES 2015 you can also `import` then with JSPM and your own compiler will compile them down to ES 5.

At this point thats probably enough. The last step would be a dev environment with `watchify`. AMD can be tackled separately from this.

**Comment by ngoldman** [comment]
@patrickarlt I don't think a full ES6/2015/? rewrite is necessary to accomplish the majority of the build tooling changes you mentioned above. My recommendation is still to let Leaflet lead in terms of conventions (if anything, propose an ES6 rewrite there and then do it here once that's happened).

With that said I'm just a bystander and occasional user, not a core contributor. If it makes maintaining and distributing easier, doesn't negatively impact contributors, and isn't a huge departure from Leaflet conventions, then I say follow your heart. :heart:

**Comment by patrickarlt** [comment]
@ngoldman its not mostly just switching up to use the ES 2015 module syntax. But I can also see a ton of uses for arrow functions `() => {}` so I can quit losing time to binding every single function to `this` and some other general simplifications.

Longer term this could also mean breaking out things like the [tiny binary search index](https://github.com/Esri/esri-leaflet/blob/master/src/Layers/FeatureLayer/FeatureManager.js#L456-L514) (used for client side time filtering) or the [conversion utils](https://github.com/Esri/esri-leaflet/blob/master/src/Util.js) (i.e. the stuff i copied from Terraformer) into separate repos. People have been asking for a bundled Terraformer + ArcGIS parser forever.

**Comment by patrickarlt** [comment]
I have an experimental 1st draft of this working on the [ES6 branch](https://github.com/Esri/esri-leaflet/tree/es6).

###### Pros

* Massively simplified build system
* 1st class support for CJS/ES 6 based projects (Browserify, WebPack, JSPM)
* Better support for AMD (load individual modules)
* Still expose a global

###### Cons

* Using ES 6 modules via Babel requires an `Object.defineProperty` polyfill for IE 8
* All of the module wrapping adds 4k (gzipped) to the file size
* Complicates testing infrastructure
* Complicates development workflow though could probably leverage JSPM to load/compile on the fly

I'm not entirely sure this is worth the benefits. This would also be a breaking change and require a major version bump so I'm thinking about doing it for 2.0.0.

**Comment by jgravois** [comment]
implemented in [2.0.0-beta.1](https://github.com/Esri/esri-leaflet/releases/tag/v2.0.0-beta.1).
